### DISCUSSION TITLE
ch2-cgo/ch2-01-hello-cgo

### DISCUSSION THREAD

**Discussion by giscus** [chai2010/advanced-go-programming-book#599] [General]
# ch2-cgo/ch2-01-hello-cgo

Go语言高级编程

https://chai2010.cn/advanced-go-programming-book/ch2-cgo/ch2-01-hello-cgo.html

**Comment by Yz-Crazy** [comment]
给哥哥们避个坑，C的代码一定要和import 在一起贴着，中间不能有空行，如果有空行会找不到C的函数
比如
```
/*
#include<stdio.h>
void printint(){
    printf("Hello World!");
}
*/

import "C" 

func main(){
    C.printint()
}
```

上面这个例子就是一个错误的例子，会报错
could not determine kind of name for C.printint
这个错误不容易发现

**Reply by lei-cheung** [reply]
你在写import "C"的时候idea会进行着色提示的，一般就能看出来的😅😅😅

**Comment by wlp7662973** [comment]
Windows系统中, 传中文的情形下, 是乱码的, 能解决么?

**Reply by PaiGack** [reply]
base64 编码，再解码出来

**Comment by PaiGack** [comment]
思考题：
```go
package main

import (
	// _GoString_ 预定义的 C 语言类型
	// 注释要和主体部分分开

	// void SayHello(_GoString_ s);
	"C"
	"fmt"

	"runtime"
	"strconv"
	"strings"
)

func GetGoid() int64 {
	var (
		buf [64]byte
		n   = runtime.Stack(buf[:], false)
		stk = strings.TrimPrefix(string(buf[:n]), "goroutine")
	)

	idField := strings.Fields(stk)[0]
	id, err := strconv.Atoi(idField)
	if err != nil {
		panic(fmt.Errorf("can not get goroutine id: %v", err))
	}

	return int64(id)
}

func main() {
	fmt.Printf("main %d\n", GetGoid())
	C.SayHello("Hello, World\n")
	fmt.Printf("main %d\n", GetGoid())
}

// export 紧跟 // 不能有空格
//
//export SayHello
func SayHello(s string) {
	fmt.Printf("SayHello %d\n", GetGoid())
	fmt.Print(s)
	fmt.Printf("SayHello %d\n", GetGoid())
}
```


```bash
go run main.go
main 1
SayHello 1
Hello, World
SayHello 1
main 1
```

**Comment by Skyfacon** [comment]
// +build go1.10

package main

//void SayHello(_GoString_ s);
import "C"

import (
    "fmt"
)

func main() {
    C.SayHello("Hello, World\n")
}

//export SayHello
func SayHello(s string) {
    fmt.Print(s)
}

**Comment by Skyfacon** [comment]
// +build go1.10

package main

//void SayHello(_GoString_ s);
import "C"

import (
    "fmt"
)

func main() {
    C.SayHello("Hello, World\n")
}

//export SayHello
func SayHello(s string) {
    fmt.Print(s)
}
我这边用这种方式，输出的是乱码呢...这是为啥呢？

**Reply by Skyfacon** [reply]
=== RUN   Test_hello
=== RUN   Test_hello/test01
--- PASS: Test_hello (0.00s)
    --- PASS: Test_hello/test01 (0.00s)
PASS

K	�f�

Process finished with the exit code 0
